### ISSUE TITLE
iFOD2: Reduce arc length on calibration failure

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#354]
In playing around with iFOD2 with the newest iteration of the SIFT phantom (in the testing repository), I found that calibrator failures could occur when the start point is right at one end of one of the bundles. The streamline would therefore fail to exit the tracking mask, or fail to reach GM in the case of ACT. This may be happening to some extent near the GM-WM interface in in vivo data.

It should theoretically be possible when this occurs to progressively reduce the number of points in the upcoming arc, both for calibration and tracking; but do not let the streamline progress any further beyond that sub-arc. Would need to adjust the FOD power accordingly also.

**Comment by jdtournier** [comment]
Just trying to get my head around what you're saying... How does the calibration failure manifest? Does it basically return zero? Or does it produce too low an amplitude so that the subsequent sampling produces a probability value higher than that (this is what's actually classed as a 'calibration failure' in the debug output)? I'm assuming the former?

In which case, the problem is that sample points at the end of the candidate arc are below threshold or have negative FOD amplitude? So the problem really is that the tracking has hit a conventional termination criterion before hitting any of the ACT ROI targets? And this is a problem specifically for the SIFT phantom you've set up for testing, or does the same thing happen on real data?

Sorry, lots of questions, but just trying to make sure I don't get the wrong idea...

**Comment by Lestropie** [comment]
What's happening specifically here is that the final point on the previous arc is _just_ within the tracking mask. The algorithm therefore attempts to continue; but all of the calibration paths return a zero probability, because at their extremities they're sampling a full half-voxel beyond the edge of the mask, where the FOD amplitudes after interpolation tend toward zero. So I'm not referring to a rejection sampling truncation error, but the calibrator reporting 'no valid paths'.

This is an extreme case because the phantom bundles end with a sharp edge. Though the test works just as well on the phantom without ACT: the tracking mask is the same as that used to specify the FODs, so while you'd expect 100% of streamlines to terminate due to exiting the mask (and this is what happens with iFOD1), you instead get ~25% terminating 'within the bundle' (though they're just shy of exiting the mask).

Won't know the prevalence on real data until I implement the suggestion and test it. But in the ACT paper, when ACT is _not_ used there's definitely a decent number terminations close to the GM-WM interface but within WM; with ACT back-tracking may help in some of these cases, but if an arc finishes just shy of the GM-WM interface, the next arc could be sampling a full 1.25mm deeper into GM signal, with noisy FODs that could plausibly be sub-threshold in the precise directions of the calibrator paths (which leads back to my last comment in 531f028).

**Comment by jdtournier** [comment]
Yes, that's what I thought you meant. Not an easy one to deal with... Give me a minute to think about this one.
